File: app/src/main/java/com/lasercyber/lws/ui/network/http/local/MonitorAlertsSseHub.java
package com.lasercyber.lws.ui.network.http.local;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.lasercyber.lws.ui.bean.entity.WarnTable;
import com.lasercyber.lws.ui.bean.event.WarnLogChangedEvent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Supplier;

/**
 * Fan-out Server-Sent Events for {@code GET /v1/monitor/alerts}.
 */
public final class MonitorAlertsSseHub {

    private static final long HEARTBEAT_INTERVAL_MS = 15_000L;
    private static final int QUEUE_CAPACITY = 64;

    private final Object lock = new Object();
    private final List<SseSubscriber> subscribers = new ArrayList<>();
    private final Supplier<List<WarnTable>> warnListSupplier;
    private final ScheduledExecutorService heartbeatExecutor;
    private final boolean autoSchedule;
    @Nullable
    private ScheduledFuture<?> heartbeatFuture;
    private long lastHeartbeatAtMs = -1L;
    private boolean eventBusRegistered;

    public MonitorAlertsSseHub(@NonNull Supplier<List<WarnTable>> warnListSupplier) {
        this(warnListSupplier, createDefaultHeartbeatExecutor(), true);
    }

    @VisibleForTesting
    MonitorAlertsSseHub(@NonNull Supplier<List<WarnTable>> warnListSupplier,
                        @NonNull ScheduledExecutorService heartbeatExecutor) {
        this(warnListSupplier, heartbeatExecutor, false);
    }

    private MonitorAlertsSseHub(@NonNull Supplier<List<WarnTable>> warnListSupplier,
                                @NonNull ScheduledExecutorService heartbeatExecutor,
                                boolean autoSchedule) {
        this.warnListSupplier = warnListSupplier;
        this.heartbeatExecutor = heartbeatExecutor;
        this.autoSchedule = autoSchedule;
    }

    public int getSubscriberCount() {
        synchronized (lock) {
            return subscribers.size();
        }
    }

    @NonNull
    public SseSubscriber acquireSubscriber() {
        synchronized (lock) {
            ensureEventBusRegisteredLocked();
            SseSubscriber sub = new SseSubscriber(this);
            subscribers.add(sub);
            startHeartbeatLocked();
            emitListLocked(sub);
            return sub;
        }
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onWarnLogChanged(@NonNull WarnLogChangedEvent event) {
        dispatchWarnEvent(event);
    }

    @VisibleForTesting
    void dispatchWarnEvent(@NonNull WarnLogChangedEvent event) {
        if (event.getKind() == WarnLogChangedEvent.Kind.CLEARED) {
            byte[] frame = encodeEvent("clear", MonitorAlertsSseJson.clearData());
            fanOut(frame);
            return;
        }
        List<WarnTable> inserted = event.getInsertedRows();
        if (inserted == null || inserted.isEmpty()) {
            return;
        }
        for (WarnTable row : inserted) {
            if (row.getId() == null) {
                continue;
            }
            byte[] frame = encodeEvent("new", MonitorAlertsSseJson.newData(row));
            fanOut(frame);
        }
    }

    private void fanOut(@NonNull byte[] frame) {
        synchronized (lock) {
            if (subscribers.isEmpty()) {
                return;
            }
            for (SseSubscriber sub : subscribers) {
                sub.offer(frame);
            }
        }
    }

    void releaseSubscriber(@NonNull SseSubscriber subscriber) {
        synchronized (lock) {
            if (!subscribers.remove(subscriber)) {
                return;
            }
            subscriber.closeInternal();
            if (subscribers.isEmpty()) {
                stopHeartbeatLocked();
            }
        }
    }

    @VisibleForTesting
    void heartbeatTickForTest() {
        heartbeatTick(System.currentTimeMillis());
    }

    @VisibleForTesting
    void heartbeatTickAtForTest(long nowMs) {
        heartbeatTick(nowMs);
    }

    @VisibleForTesting
    void resetForTest() {
        synchronized (lock) {
            stopHeartbeatLocked();
            lastHeartbeatAtMs = -1L;
            for (SseSubscriber sub : subscribers) {
                sub.closeInternal();
            }
            subscribers.clear();
            if (eventBusRegistered) {
                EventBus.getDefault().unregister(this);
                eventBusRegistered = false;
            }
        }
    }

    private void ensureEventBusRegisteredLocked() {
        if (!eventBusRegistered) {
            EventBus.getDefault().register(this);
            eventBusRegistered = true;
        }
    }

    private void emitListLocked(@NonNull SseSubscriber sub) {
        List<WarnTable> warns = warnListSupplier.get();
        sub.offer(encodeEvent("list", MonitorAlertsSseJson.listData(warns)));
    }

    private void startHeartbeatLocked() {
        if (!autoSchedule || (heartbeatFuture != null && !heartbeatFuture.isCancelled())) {
            return;
        }
        heartbeatFuture = heartbeatExecutor.scheduleAtFixedRate(
                () -> heartbeatTick(System.currentTimeMillis()),
                0L,
                HEARTBEAT_INTERVAL_MS,
                TimeUnit.MILLISECONDS);
    }

    private void stopHeartbeatLocked() {
        if (heartbeatFuture != null) {
            heartbeatFuture.cancel(false);
            heartbeatFuture = null;
        }
    }

    private void heartbeatTick(long nowMs) {
        byte[] heartbeatFrame = null;
        synchronized (lock) {
            if (subscribers.isEmpty()) {
                return;
            }
            boolean heartbeatDue = lastHeartbeatAtMs < 0L
                    || nowMs - lastHeartbeatAtMs >= HEARTBEAT_INTERVAL_MS;
            if (heartbeatDue) {
                lastHeartbeatAtMs = nowMs;
                heartbeatFrame = encodeEvent("heartbeat", MonitorAlertsSseJson.heartbeatData());
            }
            if (heartbeatFrame == null) {
                return;
            }
            for (SseSubscriber sub : subscribers) {
                sub.offer(heartbeatFrame);
            }
        }
    }

    @NonNull
    private static ScheduledExecutorService createDefaultHeartbeatExecutor() {
        return Executors.newSingleThreadScheduledExecutor(r -> {
            Thread thread = new Thread(r, "monitor-alerts-sse");
            thread.setDaemon(true);
            return thread;
        });
    }

    @NonNull
    static byte[] encodeEvent(@NonNull String event, @NonNull String jsonData) {
        String payload = "event: " + event + "\n"
                + "data: " + jsonData + "\n\n";
        return payload.getBytes(StandardCharsets.UTF_8);
    }

    public static final class SseSubscriber implements SseFrameSubscriber {
        private final MonitorAlertsSseHub hub;
        private final BlockingQueue<byte[]> queue = new LinkedBlockingQueue<>(QUEUE_CAPACITY);
        private final AtomicBoolean closed = new AtomicBoolean(false);

        SseSubscriber(@NonNull MonitorAlertsSseHub hub) {
            this.hub = hub;
        }

        void offer(@NonNull byte[] frame) {
            if (closed.get()) {
                return;
            }
            if (!queue.offer(frame)) {
                queue.poll();
                if (!queue.offer(frame)) {
                    closeFromClient();
                }
            }
        }

        @Override
        public boolean isClosed() {
            return closed.get();
        }

        @Override
        @Nullable
        public byte[] pollChunk(long timeoutMs) throws InterruptedException {
            if (closed.get() && queue.isEmpty()) {
                return null;
            }
            return queue.poll(timeoutMs, TimeUnit.MILLISECONDS);
        }

        void closeInternal() {
            closed.set(true);
            queue.offer(new byte[0]);
        }

        @Override
        public void closeFromClient() {
            if (closed.compareAndSet(false, true)) {
                queue.offer(new byte[0]);
                hub.releaseSubscriber(this);
            }
        }
    }
}

File: app/lws_hmi/lib/platform/local_http/monitor_alerts_sse_hub.dart
import 'dart:async';
import 'dart:convert';

/// Fan-out SSE for `GET /v1/monitor/alerts`.
final class MonitorAlertsSseHub {
  MonitorAlertsSseHub({
    this.listSupplier,
    this.heartbeatInterval = const Duration(seconds: 15),
  });

  /// Current WarnTable-shaped rows for immediate `list` (newest-first, ≤10).
  Future<List<Object?>> Function()? listSupplier;

  final Duration heartbeatInterval;

  static const queueCapacity = 64;
  static const listLimit = 10;

  final _subscribers = <MonitorAlertsSseSubscriber>[];
  Timer? _heartbeatTimer;
  DateTime? _lastHeartbeatAt;

  int get subscriberCount => _subscribers.length;

  Future<MonitorAlertsSseSubscriber> acquire() async {
    final sub = MonitorAlertsSseSubscriber._(this);
    _subscribers.add(sub);
    _ensureHeartbeat();
    final rows = listSupplier == null
        ? const <Object?>[]
        : await listSupplier!();
    sub.offer(_encodeEvent('list', jsonEncode(rows)));
    return sub;
  }

  void publishNew(Map<String, Object?> warnTable) {
    if (warnTable['id'] == null) {
      return;
    }
    final frame = _encodeEvent('new', jsonEncode(warnTable));
    _fanOut(frame);
  }

  void publishClear() {
    _fanOut(_encodeEvent('clear', '{}'));
  }

  void release(MonitorAlertsSseSubscriber subscriber) {
    if (!_subscribers.remove(subscriber)) {
      return;
    }
    subscriber._closeInternal();
    if (_subscribers.isEmpty) {
      _heartbeatTimer?.cancel();
      _heartbeatTimer = null;
      _lastHeartbeatAt = null;
    }
  }

  void resetForTest() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
    _lastHeartbeatAt = null;
    for (final sub in List<MonitorAlertsSseSubscriber>.from(_subscribers)) {
      sub._closeInternal();
    }
    _subscribers.clear();
  }

  void _fanOut(List<int> frame) {
    if (_subscribers.isEmpty) {
      return;
    }
    for (final sub in List<MonitorAlertsSseSubscriber>.from(_subscribers)) {
      sub.offer(frame);
    }
  }

  void _ensureHeartbeat() {
    if (_heartbeatTimer != null) {
      return;
    }
    _heartbeatTick();
    _heartbeatTimer = Timer.periodic(heartbeatInterval, (_) => _heartbeatTick());
  }

  void _heartbeatTick() {
    if (_subscribers.isEmpty) {
      return;
    }
    final now = DateTime.now();
    final due = _lastHeartbeatAt == null ||
        now.difference(_lastHeartbeatAt!) >= heartbeatInterval;
    if (!due) {
      return;
    }
    _lastHeartbeatAt = now;
    _fanOut(_encodeEvent('heartbeat', '{"ok":true}'));
  }

  void heartbeatTickForTest([DateTime? now]) {
    if (_subscribers.isEmpty) {
      return;
    }
    final at = now ?? DateTime.now();
    final due = _lastHeartbeatAt == null ||
        at.difference(_lastHeartbeatAt!) >= heartbeatInterval;
    if (!due) {
      return;
    }
    _lastHeartbeatAt = at;
    _fanOut(_encodeEvent('heartbeat', '{"ok":true}'));
  }

  static String encodeEvent(String event, String jsonData) =>
      'event: $event\ndata: $jsonData\n\n';

  static List<int> _encodeEvent(String event, String jsonData) =>
      utf8.encode(encodeEvent(event, jsonData));
}

final class MonitorAlertsSseSubscriber {
  MonitorAlertsSseSubscriber._(this._hub);

  final MonitorAlertsSseHub _hub;
  final _queue = StreamController<List<int>>.broadcast(sync: true);
  final _pending = <List<int>>[];
  bool _closed = false;

  bool get isClosed => _closed;

  Stream<List<int>> get frames async* {
    while (_pending.isNotEmpty) {
      yield _pending.removeAt(0);
    }
    yield* _queue.stream;
  }

  void offer(List<int> frame) {
    if (_closed) {
      return;
    }
    if (_queue.hasListener) {
      _queue.add(frame);
      return;
    }
    if (_pending.length >= MonitorAlertsSseHub.queueCapacity) {
      _pending.removeAt(0);
    }
    _pending.add(frame);
  }

  void _closeInternal() {
    _closed = true;
    if (!_queue.isClosed) {
      unawaited(_queue.close());
    }
  }

  void closeFromClient() {
    if (_closed) {
      return;
    }
    _hub.release(this);
  }
}
